Replace oro

FSchumacher opened the pull request:
## Description
Try to replace usage of Oro Regex implementation by Java Regex implementation. This is done by adding a property to let the user switch between the two. Default is to use the old Oro based one.

This is not finished. There are not all occurrences enhanced with the switch. A lot of it feels like a hack, as much code has been duplicated.

I thought about adding some interfaces to hide the Oro and Java Regex implementations behind, but currently I believe it would be overengineered. But maybe someone has a good idea to make this code less hacky. 

## Motivation and Context
The Oro library has been unmaintained for years and is feature-wise behind the built-in Java Regex implementation. To make a smooth transition from Oro to the Java built-in implementation the plan is to 
* make switching possible while defaulting to the old one
* deprecate the usage of the old one and switching the default to the new one
* removing the old implementation

Every step should be made with at least one release in between.
 
Discussed in [Bug 57672](https://bz.apache.org/bugzilla/show_bug.cgi?id=57672) and [Bug 65883](https://bz.apache.org/bugzilla/show_bug.cgi?id=65883)
## How Has This Been Tested?
Tests were run with the default setting (Oro Regex) and some manual tests with the setting `jmeter.use_java_regex` to `true`.
## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Delete as appropriate -->
- New feature (non-breaking change which adds functionality)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the [code style][style-guide] of this project.
- [ ] I have updated the documentation accordingly.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

FSchumacher committed:
Use isEmpty instead of comparing length to zero

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher committed:
Use isEmpty instead of comparing length to zero

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher committed:
Use isEmpty instead of comparing length to zero

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher committed:
Remove unnecessary second argument for substring (we want it all)

FSchumacher committed:
Use isEmpty instead of comparing length to zero

FSchumacher committed:
Remove unused local variable

this is the old default else branch, where implicitly  is true.

FSchumacher committed:
Use isEmpty instead of comparing length to zero

FSchumacher committed:
Add a switch to decide which regex engine should be used

pmouawad commented on src/components/src/main/java/org/apache/jmeter/assertions/JSONPathAssertion.java:
```diff
@@ -164,8 +166,12 @@ private boolean arrayMatched(JSONArray value) {
+            if (useJavaRegex) {
+                return java.util.regex.Pattern.matches(getExpectedValue(), str);
+            } else {
+                Pattern pattern = JMeterUtils.getPatternCache().getPattern(getExpectedValue());
```
I think we should also have a Cache for java Patterns as compilation has a performance penalty no ? 
Could we have a kind of factory which would give the implementation based on useJavaRegex ?

FSchumacher replied:
I can add a factory to JMeterUtils like the old one. We include caffeine, so a cache should be added real easy. For starting I chose not to add it, as it is more work and I wanted to have a working state first.

FSchumacher replied:
Added a cache with [1d30005](https://github.com/apache/jmeter/pull/700/commits/1d3000597810a7e5c379cabad00c1369002a0c1f) Now, we should use it and measure, if it is performing well :)

FSchumacher replied:
I have done some simple performance tests with the ResponseAssertion.
95% response time/99% response time/Transactions per second for *X* number of threads.

|Threads|Oro|Java (no cache)|Java (with cache)|
|-----------|-----|---------------------|-----------------------|
|1     |1.0/1.0/4266.62|1.0/1.0/5763.90|1.0/1.0/5722.43|
|10   |1.0/5.0/12939.44|1.0/5.0/12731.02|1.0/5.0/12939.44|
|100 |6.0/58.99/9862.14|6.0/77.0/11897.78|5.0/79.0/11887.85|
|1000|9.0/259.0/7867.21|3.0/56.0/9773.83|4.0/36.0/10056.35|

The test was one JSR-223 Groovy Sampler generating a long text and one ResponseAssertion asserting a contains (of the last line).

pmouawad replied:
Nice !

pmouawad reviewed (changes requested):
Hello @FSchumacher ,

Thanks for this PR which will be a great addition to JMeter.

It looks clean to me, but I think we need to have a PatternCache otherwise we will degrade JMeter performance.
Also if we could centralize the Regex implementation selection through a factory it would avoid having the property spread in multiple classes.

Thanks again for you work

vlsi commented:
Just wondering: does it replace all `oro` usages? Frankly speaking, I was surprised there are many usages of oro.

---

> `jmeter.use_java_regex`

What do you think if the property is something like `jmeter.regex.engine=oro|java` ?
Then it would be more-or-less compatible with other alternatives (if needed :) ).

FSchumacher commented:
> Just wondering: does it replace all `oro` usages? Frankly speaking, I was surprised there are many usages of oro.

```console
$ find src -name "*.java" -exec grep -rl \.oro\. {} \;
src/components/src/main/java/org/apache/jmeter/visualizers/RenderAsRegexp.java
src/components/src/main/java/org/apache/jmeter/assertions/ResponseAssertion.java
src/components/src/main/java/org/apache/jmeter/assertions/JSONPathAssertion.java
src/components/src/main/java/org/apache/jmeter/assertions/jmespath/JMESPathAssertion.java
src/components/src/main/java/org/apache/jmeter/assertions/CompareAssertion.java
src/components/src/main/java/org/apache/jmeter/extractor/RegexExtractor.java
src/core/src/test/java/org/apache/jmeter/report/dashboard/ApdexPerTransactionTest.java
src/core/src/main/java/org/apache/jmeter/engine/util/ReplaceFunctionsWithStrings.java
src/core/src/main/java/org/apache/jmeter/util/JMeterUtils.java
src/core/src/main/java/org/apache/jmeter/save/CSVSaveService.java
src/core/src/main/java/org/apache/jmeter/report/dashboard/ReportGenerator.java
src/protocol/http/src/test/java/org/apache/jmeter/protocol/http/sampler/TestHTTPSamplersAgainstHttpMirrorServer.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/accesslog/LogFilter.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/accesslog/SessionFilter.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/EncoderCache.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/parser/HtmlParsingUtils.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/parser/RegexpHTMLParser.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/modifier/URLRewritingModifier.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/config/MultipartUrlConfig.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/control/HttpMirrorThread.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/ProxyControl.java
src/functions/src/main/java/org/apache/jmeter/functions/RegexFunction.java
src/functions/src/main/java/org/apache/jmeter/functions/EscapeOroRegexpChars.java
```
I have started with the functions and assertions. The other stuff is less user visible, but would need to replaced, too.

> 
> > `jmeter.use_java_regex`
> 
> What do you think if the property is something like `jmeter.regex.engine=oro|java` ? Then it would be more-or-less compatible with other alternatives (if needed :) ).

For the bugfix, I concentrated on a feature flag to ease the transition from Oro to Java based. If we would like to be more flexible, the simplistic approach -- I took -- might not be the best and we should take more care to introduce a real abstraction instead of the if clauses.

FSchumacher committed:
Add a switch to decide which regex engine should be used

vlsi commented:
> If we would like to be more flexible, the simplistic approach

`jmeter.regex.engine=oro|java` does not require changing the implementation.

What I mean is primitive values, especially `boolean` are quite often a poor choice for an API.
In many cases, it is hard to understand what `true` and `false` means.

See https://twitter.com/lukaseder/status/1443527683720286209

On the other hand, if the only values are `oro` and `java`, then it is way easier to understand what does the switch flip.

FSchumacher commented:
> > If we would like to be more flexible, the simplistic approach
> 
> `jmeter.regex.engine=oro|java` does not require changing the implementation.
> 
> What I mean is primitive values, especially `boolean` are quite often a poor choice for an API. In many cases, it is hard to understand what `true` and `false` means.
> 
> See https://twitter.com/lukaseder/status/1443527683720286209
> 
> On the other hand, if the only values are `oro` and `java`, then it is way easier to understand what does the switch flip.

I am not really against using a more flexible approach, but I will first try to finish the classes left :)

Giving the users more choice, will let more room for misspellings, though :)

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher commented:
The `EncoderCache` from o.a.j.p.http.util uses the Cache from Oro and could probably changed to use Caffeine instead. But I will not do that in this PR.

FSchumacher committed:
Use isEmpty instead of comparing length to zero

FSchumacher committed:
Use isEmpty instead of comparing length to zero

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher committed:
Use isEmpty instead of comparing length to zero

FSchumacher committed:
Remove unused parameter in private method

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher committed:
Use isEmpty instead of comparing length to zero

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher committed:
Add a cache to store compiled Patterns into JMeterUtils

FSchumacher committed:
Remove useless semicolon

FSchumacher commented:
> `jmeter.regex.patterncache.size` would be consistent with `jmeter.regex.engine=oro|java`

Fun fact, that was my first internal name, that I changed to the underscore variant, as I didn't thought of your other remark :)

FSchumacher committed:
Rename property for regex cache size

That way -- as Vladimir pointed out -- it would be more consistent to
a potentially renamed property jmeter.regex.engine, which would
replace the boolean property jmeter.use_java_regex.

FSchumacher committed:
use 'modern' Java features as isEmpty and contains

FSchumacher committed:
Add a switch to decide which regex engine should be used

FSchumacher committed:
Fix missing 'contains'/'find' for pattern

FSchumacher commented:
URLRewritingModifier currently fails with
```
FAILURE   0,0sec, org.apache.jmeter.protocol.http.modifier.TestURLRewritingModifier > testGrabSessionId4()
    org.junit.ComparisonFailure: expected:<index.html;%24sid%24[KQNq3AAADQZoEQAxlkX8uQV5bjqVBPbT]> but was:<index.html;%24sid%24[]>
        at org.junit.Assert.assertEquals(Assert.java:117)
        at org.junit.Assert.assertEquals(Assert.java:146)
        at org.apache.jmeter.protocol.http.modifier.TestURLRewritingModifier.testGrabSessionId4(TestURLRewritingModifier.java:233)

FAILURE   0,0sec, org.apache.jmeter.protocol.http.modifier.TestURLRewritingModifier > testBug61314()
    org.junit.ComparisonFailure: expected:<...te/index;jsessionid=[]123456789> but was:<...te/index;jsessionid=[=]123456789>
        at org.junit.Assert.assertEquals(Assert.java:117)
        at org.junit.Assert.assertEquals(Assert.java:146)
        at org.apache.jmeter.protocol.http.modifier.TestURLRewritingModifier.testBug61314(TestURLRewritingModifier.java:188)

FAILURE   0,0sec, org.apache.jmeter.protocol.http.modifier.TestURLRewritingModifier > testBug61314WithQuestionMark()
    org.junit.ComparisonFailure: expected:<...te/index;jsessionid=[]123456789> but was:<...te/index;jsessionid=[=]123456789>
        at org.junit.Assert.assertEquals(Assert.java:117)
        at org.junit.Assert.assertEquals(Assert.java:146)
        at org.apache.jmeter.protocol.http.modifier.TestURLRewritingModifier.testBug61314WithQuestionMark(TestURLRewritingModifier.java:204)

FAILURE   0,0sec, org.apache.jmeter.protocol.http.modifier.TestURLRewritingModifier > testGrabSessionIdFromXMLPatExtension()
    org.junit.ComparisonFailure: expected:<...dex.html;jsessionid=[]123456789> but was:<...dex.html;jsessionid=[=]123456789>
        at org.junit.Assert.assertEquals(Assert.java:117)
        at org.junit.Assert.assertEquals(Assert.java:146)
        at org.apache.jmeter.protocol.http.modifier.TestURLRewritingModifier.testGrabSessionIdFromXMLPatExtension(TestURLRewritingModifier.java:172)

FAILURE   0,0sec,   17 completed,   4 failed,   0 skipped, org.apache.jmeter.protocol.http.modifier.TestURLRewritingModifier
```

FSchumacher committed:
Use correct order of extractors

too many No's there :)

FSchumacher commented:
> URLRewritingModifier currently fails with
> 
> ```
> ...
> FAILURE   0,0sec,   17 completed,   4 failed,   0 skipped, org.apache.jmeter.protocol.http.modifier.TestURLRewritingModifier
> ```

Found the mistake.

FSchumacher committed:
Add missing closing parenthesis in javadoc

FSchumacher committed:
Add a switch to decide which regex engine should be used

This one uses the new cache for Java based Regex-Patterns in JMeterUtils

FSchumacher committed:
Make errorProne happy and remove unused parameter from newly introduced private method

FSchumacher committed:
Add a switch to decide which regex engine should be used

This one uses the new cache for Java based Regex-Patterns in JMeterUtils

FSchumacher committed:
Use newer jupiter Assertions

FSchumacher committed:
Use isEmpty instead of comparing length to zero

FSchumacher committed:
Add a switch to decide which regex engine should be used

This one uses the new cache for Java based Regex-Patterns in JMeterUtils

FSchumacher commented:
The following files are not done yet:
```console
$ find src -name "*.java" -exec grep -rl \.oro\. {} \; | while read i; do grep -q use_java_regex $i || echo $i; done
src/core/src/main/java/org/apache/jmeter/util/JMeterUtils.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/accesslog/LogFilter.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/accesslog/SessionFilter.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/EncoderCache.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/parser/HtmlParsingUtils.java
src/functions/src/main/java/org/apache/jmeter/functions/EscapeOroRegexpChars.java
```
JMeterUtils, EncoderCache and EscapeOroRegexpChars are not problematic, as they contain no (real) regex  stuff.

LogFilter and SessionFilter have the Oro Regex API in their public API, which would be hard to remove. Should we add new methods and deprecate the old ones?

HtmlParsingUtils is a real bummer as it has the Oro stuff in its public API and it is used in our code base:
```console
$ find src -name "*.java" -exec grep -rl HtmlParsingUtils {} \; | while read i; do grep -q use_java_regex $i || echo $i; done
src/protocol/http/src/test/java/org/apache/jmeter/protocol/http/parser/TestHtmlParsingUtils.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/parser/JTidyHTMLParser.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/parser/HtmlParsingUtils.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/parser/LagartoBasedHtmlParser.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/parser/JsoupBasedHtmlParser.java
src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/modifier/AnchorModifier.java
```

vlsi commented:
> HtmlParsingUtils is a real bummer

I do not think it is.

GitHub search shows that HtmlParsingUtils is used in JMeter forks only, so there should be absolutely no issue with modifying the class: https://github.com/search?l=Java&p=1&q=org.apache.jmeter.protocol.http.parser.HtmlParsingUtils&type=Code

FSchumacher committed:
Add a switch to decide which regex engine should be used

This one uses the new cache for Java based Regex-Patterns in JMeterUtils

FSchumacher commented:
> > HtmlParsingUtils is a real bummer
> 
> I do not think it is.
> 
> GitHub search shows that HtmlParsingUtils is used in JMeter forks only, so there should be absolutely no issue with modifying the class: https://github.com/search?l=Java&p=1&q=org.apache.jmeter.protocol.http.parser.HtmlParsingUtils&type=Code

After looking into this a bit deeper, I realize, that it is not a big deal. The interface will have to be changed in the future, when we want to deprecate or drop Oro, but the internal working can be changed without a problem and the dependent classes inside JMeter used methods without Oro references.

FSchumacher commented:
With the last commit, all occurrences of Oro Regex in our codebase should be handled with a switch to change to Java Regex (apart from the few public methods, that are explicitly using Oro and will have to be deprecated in the future).

What is left:
 * Measure the impact of caching the patterns vs. compiling on the fly
 * Adding documentation
 * Changing the property name to make Vladimir happy

Should this be merged (before 5.5)?

FSchumacher committed:
Add a switch to decide which regex engine should be used

This one uses the new cache for Java based Regex-Patterns in JMeterUtils

FSchumacher committed:
Use correct variable to compile regex

FSchumacher committed:
Use private instead of protected for new methods

They should not be used outside of this class.

FSchumacher committed:
Use Pattern-cache from JMeterUtils

This will help in test plans with lots of threads (1000 tested)

FSchumacher committed:
Use Pattern cache from JMeterUtils and extract constant Pattern

FSchumacher committed:
Use Pattern cache from JMeterUtils

FSchumacher committed:
Use Pattern cache from JMeterUtils

FSchumacher committed:
Use Pattern cache from JMeterUtils

FSchumacher committed:
Use Pattern cache from JMeterUtils

FSchumacher committed:
Use Pattern cache from JMeterUtils

FSchumacher committed:
Add tests for new methods in JMeterUtils

FSchumacher committed:
Use newer junit5 Assertion methods

FSchumacher committed:
Break lines that are longer than 120 characters

FSchumacher committed:
Pass system property for regex switch to the tests

FSchumacher committed:
Java Regex and Oro differ when considering line ending

and use DOT_ALL to match new lines

FSchumacher committed:
Java Regex and oro differ in groups() vs groupCount()

Oro seems to count group(0) twice

FSchumacher committed:
Java Regex split() works quite different to oro split()

Use a different approach to split the template and correct the
difference between Java Regex groupCount() and Oro groups()

FSchumacher committed:
curly braces are special in Java Regex

Oro seems to get along with the escaped braces, so use them for testing
both implementations

FSchumacher committed:
$ is a special character in Java Regex

Guard it for usage within a Regex

FSchumacher committed:
Java Regex replaceAll() would replace too much

The Java Regex implementation sees an empty group at the end of
a string, when matching against .*. So let us rephrase our wish
to at least one character.

FSchumacher committed:
Enable testing of Java Regex instead of Oro on travis

FSchumacher committed:
Order import

FSchumacher committed:
Enable passing of system properties to tests via special namespace

If a system property prefixed with jmeter.properties. is found, Gradle will
now pass it as a system property without that prefix to the unit tests.
That allows to set system properties (which might be used as jmeter properties)
to the unit tests, without setting them in the JVM, that is used by Gradle.

pmouawad reviewed (changes requested):
Hello @FSchumacher ,
Thanks for this great PR.
I made another review.

Regards

FSchumacher committed:
Use Pattern cache in more places

FSchumacher commented on src/core/src/main/java/org/apache/jmeter/report/dashboard/ReportGenerator.java:
```diff
@@ -458,6 +458,19 @@ private ApdexSummaryConsumer createApdexSummaryConsumer() {
+            return false;
+        }
+        if (useJavaRegex) {
+            java.util.regex.Pattern pattern = java.util.regex.Pattern.compile(keyName);
```
Done

FSchumacher commented on src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/ProxyControl.java:
```diff
@@ -870,26 +871,42 @@ boolean filterContentType(SampleResult result) {
     }
 
+    private boolean isContainedWithJavaRegex(String expression, String sampleContentType) {
+        java.util.regex.Pattern pattern = java.util.regex.Pattern.compile(expression);
```
Done

FSchumacher commented on src/protocol/http/src/test/java/org/apache/jmeter/protocol/http/sampler/TestHTTPSamplersAgainstHttpMirrorServer.java:
```diff
@@ -1219,6 +1276,32 @@ private int getPositionOfBody(String stringToCheck) {
+        return getBoundaryStringFromContentTypeWithOroRegex(requestHeaders);
+    }
+
+    private String getBoundaryStringFromContentTypeWithJavaRegex(String requestHeaders) {
```
That API is not really new. Is wise to change it in this PR?

pmouawad replied:
Nothing blocking for me

vlsi commented on src/core/src/main/java/org/apache/jmeter/save/CSVSaveService.java:
```diff
@@ -557,6 +549,46 @@ public static SampleSaveConfiguration getSampleSaveConfiguration(
+                // word followed by 0 or more repeats of (non-word char + word)
+                // where the non-word char (\2) is the same
+                // e.g. abc|def|ghi but not abd|def~ghi
+                .compile("\\w+((\\W)\\w+)?(\\2\\w+)*(\\2\"\\w+\")*" // $NON-NLS-1$
```
Could you please cache it as well?

FSchumacher replied:
Yes, I can :) this escaped my grep.

FSchumacher committed:
Move static pattern to private constant

FSchumacher committed:
Use constants for static Java Regex Patterns

FSchumacher commented on src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/accesslog/SessionFilter.java:
```diff
@@ -86,6 +90,22 @@ protected boolean hasExcPattern(String text) {
+    }
+
+    private String getIpAddressWithJavaRegex(String logLine) {
+        java.util.regex.Pattern incIp = JMeterUtils.compilePattern("\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}");
```
Done (and hopefully I found all constant expressions in the new code)

FSchumacher committed:
Change property name for regex engine switch

Make switch flag internally always a privatestatic final variable, as it
might help the JIT compiler to remove dead code.

FSchumacher committed:
Add documentation for the feature flag jmeter.regex.engine

FSchumacher commented:
I think I am ready with this PR. The three things left above, are done, which were 

 * Measure the impact of caching the patterns vs. compiling on the fly
 * Adding documentation
 * Changing the property name to make Vladimir happy

Now, the question remains. Should we merge it before 5.5?
